### ISSUE TITLE
seems that client fails to start due to absence of the Mp3 directory

### ISSUE THREAD

**Issue by kambala-decapitator** [vcmi/vcmi#4973]
**Describe the bug**
Pressing Start Game on Android takes me back to the launcher

**Game logs**
[VCMI_Client_log.txt.zip](https://github.com/user-attachments/files/17893804/VCMI_Client_log.txt.zip)

Crash log:
```
Abort message: 'terminating due to uncaught exception of type std::runtime_error'
#00 0x000000000008bd08 /apex/com.android.runtime/lib64/bionic/libc.so (abort+168) (BuildId: 2f84be24a19d511a358a9050af5b3974)
#01 0x0000000000144eb8 /data/app/~~M02SvFoeXC6lh5wP4Gd6Aw==/is.xyz.vcmi.debug-I2CZcfhJ_tXVMmVrqYt6LQ==/lib/arm64/libSDL2.so (BuildId: a66e346791f5a4c8048035482b12635f5265a7ed)
                                                                                                                             abort_message
                                                                                                                             abort_message.cpp:0:0
#02 0x0000000000144b60 /data/app/~~M02SvFoeXC6lh5wP4Gd6Aw==/is.xyz.vcmi.debug-I2CZcfhJ_tXVMmVrqYt6LQ==/lib/arm64/libSDL2.so (BuildId: a66e346791f5a4c8048035482b12635f5265a7ed)
                                                                                                                             demangling_terminate_handler()
                                                                                                                             cxa_default_handlers.cpp:0:0
#03 0x0000000000144a08 /data/app/~~M02SvFoeXC6lh5wP4Gd6Aw==/is.xyz.vcmi.debug-I2CZcfhJ_tXVMmVrqYt6LQ==/lib/arm64/libSDL2.so (BuildId: a66e346791f5a4c8048035482b12635f5265a7ed)
                                                                                                                             std::__terminate(void (*)())
                                                                                                                             cxa_handlers.cpp:0:0
#04 0x000000000015f66c /data/app/~~M02SvFoeXC6lh5wP4Gd6Aw==/is.xyz.vcmi.debug-I2CZcfhJ_tXVMmVrqYt6LQ==/lib/arm64/libSDL2.so (BuildId: a66e346791f5a4c8048035482b12635f5265a7ed)
                                                                                                                             __cxxabiv1::failed_throw(__cxxabiv1::__cxa_exception*)
                                                                                                                             cxa_exception.cpp:0:0
#05 0x000000000015f5e0 /data/app/~~M02SvFoeXC6lh5wP4Gd6Aw==/is.xyz.vcmi.debug-I2CZcfhJ_tXVMmVrqYt6LQ==/lib/arm64/libSDL2.so (__cxa_throw+128) (BuildId: a66e346791f5a4c8048035482b12635f5265a7ed)
                                                                                                                             __cxa_throw
                                                                                                                             ??:0:0
#06 0x00000000027435a8 /data/app/~~M02SvFoeXC6lh5wP4Gd6Aw==/is.xyz.vcmi.debug-I2CZcfhJ_tXVMmVrqYt6LQ==/lib/arm64/libvcmiclient_arm64-v8a.so (VCMI::CFilesystemList::load(VCMI::ResourcePath const&) const+500) (BuildId: 74f8b9f7f5c84191a0856665916b07637ef1ff68)
                                                                                                                                             VCMI::CFilesystemList::load(VCMI::ResourcePath const&) const
                                                                                                                                             /Users/kambala/dev/vcmi/vcmi/lib/filesystem/AdapterLoaders.cpp:74:2
#07 0x0000000001eb7080 /data/app/~~M02SvFoeXC6lh5wP4Gd6Aw==/is.xyz.vcmi.debug-I2CZcfhJ_tXVMmVrqYt6LQ==/lib/arm64/libvcmiclient_arm64-v8a.so (BuildId: 74f8b9f7f5c84191a0856665916b07637ef1ff68)
                                                                                                                                             MusicEntry::load(VCMI::ResourcePathTempl<(VCMI::EResType)11> const&)
                                                                                                                                             /Users/kambala/dev/vcmi/vcmi/client/media/CMusicHandler.cpp:268:67
#08 0x0000000001eb6e18 /data/app/~~M02SvFoeXC6lh5wP4Gd6Aw==/is.xyz.vcmi.debug-I2CZcfhJ_tXVMmVrqYt6LQ==/lib/arm64/libvcmiclient_arm64-v8a.so (BuildId: 74f8b9f7f5c84191a0856665916b07637ef1ff68)
                                                                                                                                             MusicEntry::MusicEntry(CMusicHandler*, std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char>>, VCMI::ResourcePathTempl<(VCMI::EResType)11> const&, bool, bool)
                                                                                                                                             /Users/kambala/dev/vcmi/vcmi/client/media/CMusicHandler.cpp:224:3
#09 0x0000000001eb6a1c /data/app/~~M02SvFoeXC6lh5wP4Gd6Aw==/is.xyz.vcmi.debug-I2CZcfhJ_tXVMmVrqYt6LQ==/lib/arm64/libvcmiclient_arm64-v8a.so (BuildId: 74f8b9f7f5c84191a0856665916b07637ef1ff68)
                                                                                                                                             std::__ndk1::__unique_if<MusicEntry>::__unique_single std::__ndk1::make_unique[abi:ne180000]<MusicEntry, CMusicHandler*&, std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char>> const&, VCMI::ResourcePathTempl<(VCMI::EResType)11> const&, bool&, bool&>(CMusicHandler*&, std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char>> const&, VCMI::ResourcePathTempl<(VCMI::EResType)11> const&, bool&, bool&)
                                                                                                                                             /Users/kambala/.conan2/p/androe174383464d30/p/bin/toolchains/llvm/prebuilt/darwin-x86_64/sysroot/usr/include/c++/v1/__memory/unique_ptr.h:601:30
#10 0x0000000001eb5ed4 /data/app/~~M02SvFoeXC6lh5wP4Gd6Aw==/is.xyz.vcmi.debug-I2CZcfhJ_tXVMmVrqYt6LQ==/lib/arm64/libvcmiclient_arm64-v8a.so (BuildId: 74f8b9f7f5c84191a0856665916b07637ef1ff68)
                                                                                                                                             CMusicHandler::queueNext(CMusicHandler*, std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char>> const&, VCMI::ResourcePathTempl<(VCMI::EResType)11> const&, bool, bool)
                                                                                                                                             /Users/kambala/dev/vcmi/vcmi/client/media/CMusicHandler.cpp:152:12
#11 0x0000000001eb5d3c /data/app/~~M02SvFoeXC6lh5wP4Gd6Aw==/is.xyz.vcmi.debug-I2CZcfhJ_tXVMmVrqYt6LQ==/lib/arm64/libvcmiclient_arm64-v8a.so (BuildId: 74f8b9f7f5c84191a0856665916b07637ef1ff68)
                                                                                                                                             CMusicHandler::playMusic(VCMI::ResourcePathTempl<(VCMI::EResType)11> const&, bool, bool)
                                                                                                                                             /Users/kambala/dev/vcmi/vcmi/client/media/CMusicHandler.cpp:110:2
#12 0x0000000001dab6f4 /data/app/~~M02SvFoeXC6lh5wP4Gd6Aw==/is.xyz.vcmi.debug-I2CZcfhJ_tXVMmVrqYt6LQ==/lib/arm64/libvcmiclient_arm64-v8a.so (BuildId: 74f8b9f7f5c84191a0856665916b07637ef1ff68)
                                                                                                                                             CMainMenu::playMusic()
                                                                                                                                             /Users/kambala/dev/vcmi/vcmi/client/mainmenu/CMainMenu.cpp:330:15
#13 0x0000000001dd25e0 /data/app/~~M02SvFoeXC6lh5wP4Gd6Aw==/is.xyz.vcmi.debug-I2CZcfhJ_tXVMmVrqYt6LQ==/lib/arm64/libvcmiclient_arm64-v8a.so (BuildId: 74f8b9f7f5c84191a0856665916b07637ef1ff68)
                                                                                                                                             CMainMenu::playIntroVideos()::$_1::operator()(bool) const
                                                                                                                                             /Users/kambala/dev/vcmi/vcmi/client/mainmenu/CMainMenu.cpp:324:4
#14 0x0000000001dd24f4 /data/app/~~M02SvFoeXC6lh5wP4Gd6Aw==/is.xyz.vcmi.debug-I2CZcfhJ_tXVMmVrqYt6LQ==/lib/arm64/libvcmiclient_arm64-v8a.so (BuildId: 74f8b9f7f5c84191a0856665916b07637ef1ff68)
                                                                                                                                             decltype(std::declval<CMainMenu::playIntroVideos()::$_1&>()(std::declval<bool>())) std::__ndk1::__invoke[abi:ne180000]<CMainMenu::playIntroVideos()::$_1&, bool>(CMainMenu::playIntroVideos()::$_1&, bool&&)
                                                                                                                                             /Users/kambala/.conan2/p/androe174383464d30/p/bin/toolchains/llvm/prebuilt/darwin-x86_64/sysroot/usr/include/c++/v1/__type_traits/invoke.h:344:25
#15 0x0000000001dd249c /data/app/~~M02SvFoeXC6lh5wP4Gd6Aw==/is.xyz.vcmi.debug-I2CZcfhJ_tXVMmVrqYt6LQ==/lib/arm64/libvcmiclient_arm64-v8a.so (BuildId: 74f8b9f7f5c84191a0856665916b07637ef1ff68)
                                                                                                                                             void std::__ndk1::__invoke_void_return_wrapper<void, true>::__call[abi:ne180000]<CMainMenu::playIntroVideos()::$_1&, bool>(CMainMenu::playIntroVideos()::$_1&, bool&&)
                                                                                                                                             /Users/kambala/.conan2/p/androe174383464d30/p/bin/toolchains/llvm/prebuilt/darwin-x86_64/sysroot/usr/include/c++/v1/__type_traits/invoke.h:419:5
#16 0x0000000001dd2470 /data/app/~~M02SvFoeXC6lh5wP4Gd6Aw==/is.xyz.vcmi.debug-I2CZcfhJ_tXVMmVrqYt6LQ==/lib/arm64/libvcmiclient_arm64-v8a.so (BuildId: 74f8b9f7f5c84191a0856665916b07637ef1ff68)
                                                                                                                                             std::__ndk1::__function::__alloc_func<CMainMenu::playIntroVideos()::$_1, std::__ndk1::allocator<CMainMenu::playIntroVideos()::$_1>, void (bool)>::operator()[abi:ne180000](bool&&)
                                                                                                                                             /Users/kambala/.conan2/p/androe174383464d30/p/bin/toolchains/llvm/prebuilt/darwin-x86_64/sysroot/usr/include/c++/v1/__functional/function.h:166:12
#17 0x0000000001dd152c /data/app/~~M02SvFoeXC6lh5wP4Gd6Aw==/is.xyz.vcmi.debug-I2CZcfhJ_tXVMmVrqYt6LQ==/lib/arm64/libvcmiclient_arm64-v8a.so (BuildId: 74f8b9f7f5c84191a0856665916b07637ef1ff68)
                                                                                                                                             std::__ndk1::__function::__func<CMainMenu::playIntroVideos()::$_1, std::__ndk1::allocator<CMainMenu::playIntroVideos()::$_1>, void (bool)>::operator()(bool&&)
                                                                                                                                             /Users/kambala/.conan2/p/androe174383464d30/p/bin/toolchains/llvm/prebuilt/darwin-x86_64/sysroot/usr/include/c++/v1/__functional/function.h:308:10
#18 0x0000000001dca25c /data/app/~~M02SvFoeXC6lh5wP4Gd6Aw==/is.xyz.vcmi.debug-I2CZcfhJ_tXVMmVrqYt6LQ==/lib/arm64/libvcmiclient_arm64-v8a.so (BuildId: 74f8b9f7f5c84191a0856665916b07637ef1ff68)
                                                                                                                                             std::__ndk1::__function::__value_func<void (bool)>::operator()[abi:ne180000](bool&&) const
                                                                                                                                             /Users/kambala/.conan2/p/androe174383464d30/p/bin/toolchains/llvm/prebuilt/darwin-x86_64/sysroot/usr/include/c++/v1/__functional/function.h:425:12
#19 0x0000000001dc79cc /data/app/~~M02SvFoeXC6lh5wP4Gd6Aw==/is.xyz.vcmi.debug-I2CZcfhJ_tXVMmVrqYt6LQ==/lib/arm64/libvcmiclient_arm64-v8a.so (BuildId: 74f8b9f7f5c84191a0856665916b07637ef1ff68)
                                                                                                                                             std::__ndk1::function<void (bool)>::operator()(bool) const
                                                                                                                                             /Users/kambala/.conan2/p/androe174383464d30/p/bin/toolchains/llvm/prebuilt/darwin-x86_64/sysroot/usr/include/c++/v1/__functional/function.h:978:10
#20 0x0000000001dab5d0 /data/app/~~M02SvFoeXC6lh5wP4Gd6Aw==/is.xyz.vcmi.debug-I2CZcfhJ_tXVMmVrqYt6LQ==/lib/arm64/libvcmiclient_arm64-v8a.so (BuildId: 74f8b9f7f5c84191a0856665916b07637ef1ff68)
                                                                                                                                             CMainMenu::playIntroVideos()::$_0::operator()(std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char>>, bool, float, std::__ndk1::function<void (bool)>) const
                                                                                                                                             /Users/kambala/dev/vcmi/vcmi/client/mainmenu/CMainMenu.cpp:310:4
#21 0x0000000001dab284 /data/app/~~M02SvFoeXC6lh5wP4Gd6Aw==/is.xyz.vcmi.debug-I2CZcfhJ_tXVMmVrqYt6LQ==/lib/arm64/libvcmiclient_arm64-v8a.so (BuildId: 74f8b9f7f5c84191a0856665916b07637ef1ff68)
                                                                                                                                             CMainMenu::playIntroVideos()
                                                                                                                                             /Users/kambala/dev/vcmi/vcmi/client/mainmenu/CMainMenu.cpp:313:2
#22 0x000000000199d6cc /data/app/~~M02SvFoeXC6lh5wP4Gd6Aw==/is.xyz.vcmi.debug-I2CZcfhJ_tXVMmVrqYt6LQ==/lib/arm64/libvcmiclient_arm64-v8a.so (SDL_main+12112) (BuildId: 74f8b9f7f5c84191a0856665916b07637ef1ff68)
                                                                                                                                             SDL_main
...
```

**To Reproduce**
1. Don't add Mp3 dir
2. Start the game

**Expected behavior**
no crash

**Actual behavior**
crash

**Did it work earlier?**
Works in 1.5.7

**Version**
 - OS: Android 12
 - Version: latest develop - a4417f3fc5cbbbb56948fcfb7b62e08b9edbb9a2

**Additional context**
Dependencies and app are built with NDK r27 (clang 18), not sure if it might be coming from there

**Comment by IvanSavenko** [comment]
Strange. Probably somehow caused by new NDK / compiler.
std::runtime_error is thrown here:
`/Users/kambala/dev/vcmi/vcmi/lib/filesystem/AdapterLoaders.cpp:74:2`
But there is try/catch block for std::exception immediately in the caller method:
`/Users/kambala/dev/vcmi/vcmi/client/media/CMusicHandler.cpp:268:67`

**Comment by kambala-decapitator** [comment]
> But there is try/catch block for std::exception immediately in the caller method:

could be NDK quirks...

**Comment by IvanSavenko** [comment]
> @IvanSavenko could you check behavior using the build artifact?

Android one? Probably not now. Can't connect it to PC to remove Mp3 directory.

> could be NDK quirks...

Library and client are still linked statically? I do remember some quirks where instances of classes instantiated in library wont work correctly in client (since client had own version of that type with separate RTTI data).

We could also catch std::runtime_error directly instead of std::exception, or even add our own ResourceNotFoundException. But I suspect this is not the only place with such code...

**Comment by kambala-decapitator** [comment]
> Library and client are still linked statically?

yes, no project changes

> I do remember some quirks where instances of classes instantiated in library wont work correctly in client (since client had own version of that type with separate RTTI data).

I remember something like that with shared AI libs on android: exception from AI was never caught by client

> We could also catch std::runtime_error directly instead of std::exception, or even add our own ResourceNotFoundException

thanks, will try that, at least to understand if it could be caught at all

**Comment by kambala-decapitator** [comment]
catching `std::runtime_error` explicitly helped

**Comment by kambala-decapitator** [comment]
otherwise only `...` catches. Made 1 turn on a predefined map - everything was fine.

Sounds like NDK bug to me, will try to reproduce in a clean test project. However, release notes of patch releases don't mention anything like that. Will still test with r27c once it's available in Conan.

**Comment by kambala-decapitator** [comment]
build from CI works fine

**Comment by IvanSavenko** [comment]
Not sure if related, but: 
https://github.com/android/ndk/issues/889#issuecomment-460568023

Specifically:
> In one of those libs, I define a custom exception inheriting from std::exception, and when I throw a custom exception from lib1 and try to catch it from lib2 I have a crash on the symbol Unwind_Resume in libc.so.

> What I did not see directly was that the symbol before this one was an Unwind_Resume symbol in libz.so. So why libz appears in my stack trace ?

I see same in your log - why there is libSDL2.so in call stack?

**Comment by kambala-decapitator** [comment]
> why there is libSDL2.so in call stack?

that's a good question. Will build it with debug symbols to understand what's going on.

**Comment by kambala-decapitator** [comment]
regarding SDL I don't have a good answer. At first I thought it's because VCMI runs from SDL (SDL_main function), but then I placed test code into launcher and got the same crash, and that one doesn't have SDL down the stack trace. Enabling debug symbols in SDL didn't show anything because __cxa_throw is a clang/runtime function. Printing exception's type_info didn't reveal anything unusual.

What I also tested was a plain cpp program that just throws and catches - it works fine, also tried mimicking our setup a bit by throwing exception from a shared lib and catching it in a static one. A simple Android app that just launches some native code (took SDL sample) also doesn't have the issue. I also tried adding compiler flags that VCMI is built with, but that also didn't have any effect.

Test code is:
```cpp
#include <stdexcept>
#include <android/log.h>
extern "C" int SDL_main(int argc, const char* argv[]) {
	try {
		throw std::runtime_error{""};
	}
	catch(const std::exception& e)
	{
		__android_log_write(ANDROID_LOG_INFO, "qwe", "std::exception");
	}
	return 0;
}
```

**Comment by kambala-decapitator** [comment]
same issue with r26d (clang 17, and we currently use clang 14), looks like toolchain upgrade is postponed. At least now we know that something seriously changed in r26.

**Comment by kambala-decapitator** [comment]
now that's interesting! I've built everything with the currently used toolchain (r25c) and it crashes as well! But the only changes are:
- newer libs used
- Conan v2 vs v1

**Comment by kambala-decapitator** [comment]
current changes: https://github.com/vcmi/vcmi/compare/develop...kambala-decapitator:vcmi:conan-v2

**Comment by IvanSavenko** [comment]
Is it possible that some of libraries were somehow built with static libc++ and some with shared?

Is it possible that vcmi or some libraries were built without exceptions or RTTI support? NDK docs imply that they are off by default. Unlikely, but what if?

Can you upload faulty apk?

**Comment by kambala-decapitator** [comment]
everything's possible, who knows what flags a lib might enforce. VCMI is definitely built with exceptions, otherwise changing the caught exception type in code wouldn't have worked around the issue. Can try to inject rtti/exceptions flags explicitly into Conan profile.

apk: https://www.dropbox.com/scl/fi/8p4qj5vie0et70qf25hcg/vcmi-debug.apk?rlkey=i27a24uk1kcjirl1tc5rchroy&dl=0

**Comment by IvanSavenko** [comment]
libSDL is now shared. Was static before. 
Also, not 100% sure, but I think libSDL was built with static libstdc++

Compared .so in your library with what we had before:
- some new Qt .so, all need libc++_shared (not sure why we now have them, but probably safe)
- (!) we now have libSDL.so and it does not need libc++ (even though this is c++ library) 
- we also have libSDL components (image/ttf/mixer). But these looks like pure C libraries, so probably safe?
- libogg and libopus now exist, but they looks to be pure C
- libfreetype is now static? I believe it is used by both Qt and SDL_ttf, so makes sense to keep shared, but safe.

So I think the only suspicious change is libSDL. Can you try to investigate why it uses static libc++? Or build it statically as before?

**Comment by kambala-decapitator** [comment]
> some new Qt .so, all need libc++_shared (not sure why we now have them, but probably safe)

you mean some new plugin? That should be safe, but can't say why it appeared. The version is still 5.15.14.

> we also have libSDL components (image/ttf/mixer). But these looks like pure C libraries, so probably safe?

yes, I think so. Btw the recommendation is to have either all of them shared or all static (well, in general it applies to all consumed libs by a binary).

> libfreetype is now static? I believe it is used by both Qt and SDL_ttf, so makes sense to keep shared, but safe.

yes, it should be shared, but currently Conan's Qt can't be configured in such case, I opened bug about that https://github.com/conan-io/conan-center-index/issues/26020

> checked SDL source and this is C library

Yes, should be pure C.

Will try static and check where C++ might be coming from.

**Comment by IvanSavenko** [comment]
> you mean some new plugin? That should be safe, but can't say why it appeared. The version is still 5.15.14.

Ah, sorry, we actually *lost* some Qt libraries. Mixed up which side is before and which is after in my comparison.

```
libQt5Concurrent_arm64-v8a.so
libQt5Help_arm64-v8a.so
libQt5OpenGL_arm64-v8a.so
libQt5PrintSupport_arm64-v8a.so
libQt5Sql_arm64-v8a.so
libQt5Test_arm64-v8a.so
libQt5Xml_arm64-v8a.so
```
None of these exist in your build, but exist in latest daily build. And most of these sound like stuff we don't use anyway.

In any case, this is just an observation, extremely unlikely to be related to this crash.

**Comment by kambala-decapitator** [comment]
yes, I improved the Qt libs copy process to include only what's needed

**Comment by kambala-decapitator** [comment]
> But `readelf` reports references to c++-only symbols in libSDL for some reason. Including those related to exception throwing.

you're absolutely right, SDL does link to libc++ (which is also visible in the stacktrace). Mystery solved: it does so only on Android and only when hidapi subsystem is requested to be built, but Conan's recipe "deletes" libc++ unconditionally, hence default libc++ setting comes to play which is set to static in the NDK cmake toolchain. For static build it's not an issue of course because there's nothing to link. Now all works fine and I've submitted a fix: https://github.com/conan-io/conan-center-index/pull/26049

Also, it's better to make SDL static for iOS and Android (single app platforms basically), but it's good that we found this issue :)

P.S. libc++ is also linked for WinRT and Haiku platforms (all SDL platform code is written in C++), but we don't care about them.
